sentry_sdk/hub.py

```diff
@@ -441,38 +441,138 @@ def start_span(
     ):
         # type: (...) -> Span
         """
-        Create a new span whose parent span is the currently active
-        span, if any. The return value is the span object that can
-        be used as a context manager to start and stop timing.
-
-        Note that you will not see any span that is not contained
-        within a transaction. Create a transaction with
-        ``start_span(transaction="my transaction")`` if an
-        integration doesn't already do this for you.
+        Create and start timing a new span whose parent is the currently active
+        span or transaction, if any. The return value is a span instance,
+        typically used as a context manager to start and stop timing in a `with`
+        block.
+
+        Only spans contained in a transaction are sent to Sentry. Most
+        integrations start a transaction at the appropriate time, for example
+        for every incoming HTTP request. Use `start_transaction` to start a new
+        transaction when one is not already in progress.
         """
-
-        client, scope = self._stack[-1]
+        # XXX: if the only way to use start_span correctly is when there
+        # is already an existing transaction/span in the scope, then
+        # this is a hard to use API. We don't document nor support
+        # appending an existing span to a new transaction created to
+        # contain the span.
+
+        # TODO: consider removing this in a future release.
+        # This is for backwards compatibility with releases before
+        # start_transaction existed, to allow for a smoother transition.
+        if isinstance(span, Transaction) or "transaction" in kwargs:
+            deprecation_msg = (
+                "Deprecated: use start_transaction to start transactions and "
+                "Transaction.start_child to start spans."
+            )
+            if isinstance(span, Transaction):
+                logger.warning(deprecation_msg)
+                return self.start_transaction(span)
+            if "transaction" in kwargs:
+                logger.warning(deprecation_msg)
+                name = kwargs.pop("transaction")
+                return self.start_transaction(name=name, **kwargs)
+
+        # XXX: this is not very useful because
+        #
+        #     with hub.start_span(Span(...)):
+        #         pass
+        #
+        # is equivalent to the simpler
+        #
+        #     with Span(...):
+        #         pass
+        if span is not None:
+            return span
 
         kwargs.setdefault("hub", self)
 
-        if span is None:
-            span = scope.span
-            if span is not None:
-                span = span.new_span(**kwargs)
-            else:
-                span = Span(**kwargs)
+        span = self.scope.span
+        if span is not None:
+            return span.start_child(**kwargs)
+
+        # XXX: returning a detached span here means whatever span tree built
```

Comment:
This makes the API simpler to use as the caller does not have to check whether they are in a transaction. I don't really see why you call this not useful.

Reply:
This can hide an entire instrumented span tree if one is not careful to guarantee a transaction is in the scope when calling `hub.start_span(op="...")`.

You told me today about a use case for intentionally omitting an entire instrumented span tree, I still keeping wrap my head around that. In other parts of the Unified API we explicitly prefer to avoid doing things when not necessary -- e.g. no-op transport/client when SDK has no DSN, callbacks are not called, etc.

I'll move these discussions into issues.